Parameter shape placeholder

bkvogel opened the pull request:
Provide support to allow a link to adjust the sizes of its parameters based on the sizes of its inputs, potentially deferring initialization of some parameters until the first forward call.

Many existing links make use of one or more parameters that have a size that depends on the size of the input(s) to the link. An example of this is the Linear link, which has a weights parameter `W` in which one of its dimensions depends on the size of the input activations. In the existing implementation, all parameter sizes are initialized in the link initializer, which requires that the user pass this size information as an argument when creating the link. For example, in the Linear link, `in_size` must be specified as an argument to the initializer.

When creating deep networks, however, it can be inconvenient for the user to determine the required size arguments because the sizes can be altered by the parent links in the computation graph. Furthermore, such size information would not need to be specified at all if parameter initialization were deferred until the first forward call because the size information could then be directly obtained from the input(s). 

This feature adds support for parameters initialization to be deferred until the first forward call. This PR adds support the the Linear and Convolution2D links. Support for remaining links will be added as a future PR.

To use this feature, simply pass `None` as the value for the corresponding argument in the initializer. For example, to use deferred initialization of `W` in the Linear link, pass `None` as the `in_size` argument. This maintains backward compatibility.

unnonouno commented on chainer/link.py:
```diff
@@ -134,6 +140,49 @@ def add_param(self, name, shape, dtype=numpy.float32):
+            return False.
+
+        """
+        if len(self._uninitialized_params) > 0:
```
`return len(self._uninitialized_params) > 0` ?

bkvogel replied:
I will make the change.

unnonouno commented:
I feel the basic design looks good. Please add a test for serialization to `test_link.py`.

unnonouno commented on chainer/link.py:
```diff
@@ -134,6 +141,46 @@ def add_param(self, name, shape, dtype=numpy.float32):
+    def has_uninitialized_params(self):
+        """Check if the link has uninitialized parameters.
+
+        Returns: True if the link has any uninitialized parameters. Otherwise
```
Insert  a new line after `Returns:` and add type `bool:` before the description.

unnonouno commented on chainer/link.py:
```diff
@@ -134,6 +141,46 @@ def add_param(self, name, shape, dtype=numpy.float32):
+        """
+        d = self.__dict__
+        if (name in self._uninitialized_params) or (name in d):
+            raise AttributeError(
```
Test if the method raises this error when a given param is registered.

unnonouno commented on chainer/serializers/npz.py:
```diff
@@ -94,7 +94,9 @@ def __getitem__(self, key):
         key = key.lstrip('/')
         dataset = self.npz[self.path + key]
-        if isinstance(value, numpy.ndarray):
+        if value is None:
```
Can you add a test for this fix?

unnonouno replied:
Do we need to fix hdf5 serializer?

unnonouno replied:
Do we need to fix document of `AbstractSerializer`? In its docstring, `value` is assumed to be `scalar, array, or str`.

bkvogel replied:
Yes, it looks like I should add the check for `value` is `None` to `HDF5Deserializer` as well. In that case, I would agree that we should fix the documentation of `AbstractSerializer` to mention that for deserializers, `value` can also be `None`.

unnonouno commented on chainer/link.py:
```diff
@@ -134,6 +141,46 @@ def add_param(self, name, shape, dtype=numpy.float32):
+        self._uninitialized_params.add(name)
+
+    @property
+    def has_uninitialized_params(self):
```
Please add a unit test for this method in `test_link.py`.

unnonouno commented:
I wrote some comments especially about test. Is specification of serializer changed? I mean, serializers need to support `value=None`.

bkvogel commented:
Thank you for your comments. I have added the requested tests. I also added support for optional `value` of `None` in ` HDF5Deserializer` so that both deserializers behave similarly and updated documentation of base class to mention that `None` is also a possible argument value. Could you check the changes? Thanks.

unnonouno commented:
LGTM except one comment about docstring.

bkvogel commented:
Thanks for your comments. I have fixed the indent.

unnonouno commented:
Great!! Thank you.
